src/Core/Indexes/KDTree.cs

```diff
@@ -0,0 +1,724 @@
+        public bool Remove(Envelope envelope)
+        {
+            bool result = false;
+            IEnumerable<Coordinate> coordinates = this.Search(envelope);
+            foreach (Coordinate coordinate in coordinates)
+            {
+                if (this.Remove(coordinate))
+                    result = true;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Attempts to remove all coordinates inside an envelope.
+        /// </summary>
+        /// <param name="envelope">The envelope.</param>
+        /// <param name="coordinates">The coordinates which were removed.</param>
+        /// <returns>A value indicating whether any coordinates were removed.</returns>
+        /// <exception cref="System.ArgumentNullException">The envelope is null.</exception>
+        public bool Remove(Envelope envelope, out List<Coordinate> coordinates)
+        {
+            bool result = false;
+            coordinates = this.Search(envelope).ToList();
+            foreach (Coordinate coordinate in coordinates)
+            {
+                if (this.Remove(coordinate))
+                    result = true;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Searches the tree for all coordinates inside the specified envelope.
+        /// </summary>
+        /// <param name="envelope">The envelope.</param>
+        /// <returns>All coordinates from the envelope which are in the tree.</returns>
+        /// <exception cref="System.ArgumentNullException">The envelope is null.</exception>
+        public IEnumerable<Coordinate> Search(Envelope envelope)
+        {
+            if (envelope == null)
+                throw new ArgumentNullException("envelope", "The envelope is null.");
+
+            if (this.IsEmpty)
+                return new List<Coordinate>();
+
+            return this.root.Search(envelope);
+        }
+
+        /// <summary>
+        /// Searches the tree for the nearest neighbour of the specified coordinate.
+        /// </summary>
+        /// <param name="coordinate">The coordinate.</param>
+        /// <returns>The nearest neighbour of the specified coordinate.</returns>
+        /// <exception cref="System.ArgumentNullException">The coordinate is null.</exception>
+        /// <exception cref="System.InvalidOperationException">The tree is empty.</exception>
+        public Coordinate NearestNeighbourSearch(Coordinate coordinate)
+        {
+            if (coordinate == null)
+                throw new ArgumentNullException("coordinate", "The coordinate is null.");
+            if (this.IsEmpty)
+                throw new InvalidOperationException("Cannot perform nearest neighbour search on an empty tree.");
+
+            return this.root.NearestNeighbourSearch(coordinate);
+        }
+
+        /// <summary>
+        /// Rebalances the tree. The tree is balanced when created, but adding or removing elements might leave it unbalanced.
```

Comment:
Please keep the `summary` section short and add additional comments to the `remarks` section.